send-transaction-service/src/send_transaction_service.rs

```diff
@@ -703,70 +666,101 @@ impl SendTransactionService {
             let wire_transactions = transactions
                 .iter()
                 .filter(|(signature, _)| batched_transactions.contains(signature))
-                .map(|(_, transaction_info)| transaction_info.wire_transaction.as_ref())
-                .collect::<Vec<&[u8]>>();
-
-            let iter = wire_transactions.chunks(config.batch_size);
-            for chunk in iter {
-                let mut addresses = config
-                    .tpu_peers
-                    .as_ref()
-                    .map(|addrs| addrs.iter().collect::<Vec<_>>())
-                    .unwrap_or_default();
-                let mut leader_info_provider = leader_info_provider.lock().unwrap();
-                let leader_info = leader_info_provider.get_leader_info();
-                let leader_addresses = Self::get_tpu_addresses(
-                    tpu_address,
-                    leader_info,
-                    config,
-                    connection_cache.protocol(),
-                );
-                addresses.extend(leader_addresses);
-
-                for address in &addresses {
-                    Self::send_transactions(address, chunk, connection_cache, stats);
-                }
+                .map(|(_, transaction_info)| transaction_info.wire_transaction.clone());
+
+            let iter = wire_transactions.chunks(batch_size);
+            for chunk in &iter {
+                let chunk = chunk.collect();
+                client.send_transactions_in_batch(chunk, stats);
             }
         }
         result
     }
 
-    fn send_transaction(
-        tpu_address: &SocketAddr,
-        wire_transaction: &[u8],
-        connection_cache: &Arc<ConnectionCache>,
-    ) -> Result<(), TransportError> {
-        let conn = connection_cache.get_connection(tpu_address);
-        conn.send_data_async(wire_transaction.to_vec())
+    pub fn join(self) -> thread::Result<()> {
+        self.receive_txn_thread.join()?;
+        self.exit.store(true, Ordering::Relaxed);
+        self.retry_thread.join()
     }
+}
 
-    fn send_transactions_with_metrics(
-        tpu_address: &SocketAddr,
-        wire_transactions: &[&[u8]],
-        connection_cache: &Arc<ConnectionCache>,
-    ) -> Result<(), TransportError> {
-        let wire_transactions = wire_transactions.iter().map(|t| t.to_vec()).collect();
-        let conn = connection_cache.get_connection(tpu_address);
-        conn.send_data_batch_async(wire_transactions)
+pub trait TransactionClient {
```

Comment:
This client wraps all the logic related to the leader selection together with sending transactions. This effectively splits the networking code from the rest opening the route for the replacement of the network code in the follow up PRs.